src/Lykke.AlgoStore.MatchingEngineAdapter.Services/Listening/ListeningService.cs

```diff
@@ -0,0 +1,136 @@
+        private readonly ConsumerLoadBalancer _consumerLoadBalancer;
+        private readonly CancellationTokenSource _cts = new CancellationTokenSource();
+        private readonly ushort _port;
+        private readonly ILog _log;
+
+        private TcpListener _listener;
+
+        private Thread _acceptingThread;
+
+        private bool _isDisposed;
+
+        /// <summary>
+        /// Initializes a <see cref="ListeningService"/>
+        /// </summary>
+        public ListeningService(IProducerLoadBalancer producerLoadBalancer, IMessageQueue requestQueue, IMatchingEngineAdapter matchingEngineAdapter,
+            ushort port, [NotNull] ILog log)
+        {
+            _requestQueue = requestQueue ?? throw new ArgumentNullException(nameof(requestQueue));
+            _matchingEngineAdapter = matchingEngineAdapter ?? throw new ArgumentNullException(nameof(matchingEngineAdapter));
+            _log = log ?? throw new ArgumentNullException(nameof(log));
+
+            _producerLoadBalancer = producerLoadBalancer ?? throw new ArgumentNullException(nameof(producerLoadBalancer));
+            _consumerLoadBalancer = new ConsumerLoadBalancer(_requestQueue, _matchingEngineAdapter, _log);
+
+            _port = port;
+        }
+
+        /// <summary>
+        /// Starts listening and starts the connection accepting thread
+        /// </summary>
+        public void Start()
+        {
+            _listener = new TcpListener(IPAddress.Any, _port);
+            _listener.Start();
+
+            _acceptingThread = new Thread(AcceptConnections);
+            _acceptingThread.Priority = ThreadPriority.Highest;
+            _acceptingThread.Start(_cts.Token);
+        }
+
+        /// <summary>
+        /// Disposes this <see cref="ListeningService"/> by shutting down the accepting thread,
+        /// the listener and disposing the load balancers
+        /// </summary>
+        public void Stop()
+        {
+            Dispose();
+        }
+
+        /// <summary>
+        /// Disposes this <see cref="ListeningService"/> by shutting down the accepting thread,
+        /// the listener and disposing the load balancers
+        /// </summary>
+        public void Dispose()
+        {
+            Dispose(true);
+        }
+
+        protected virtual void Dispose(bool isDisposing)
+        {
+            if (_isDisposed) return;
+
+            if (isDisposing)
+            {
+                _cts.Cancel();
+                _acceptingThread.Join();
+
+                _listener.Stop();
+                _producerLoadBalancer.Dispose();
+                _consumerLoadBalancer.Dispose();
+            }
+
+            _isDisposed = true;
+        }
+
+        /// <summary>
+        /// Accepts a new connection until the thread is stopped
+        /// </summary>
+        /// <param name="cancellationTokenObj">A <see cref="CancellationToken"/></param>
+        private void AcceptConnections(object cancellationTokenObj)
+        {
```

Comment:
The method looked overcomplicated. Many disposables left without disposing. Can it be rewritten with async|await?